stout/take.h

```diff
@@ -0,0 +1,191 @@
+
+namespace stout {
+namespace eventuals {
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////
+
+struct _TakeLastN {
+  template <typename K_, typename Arg_>
+  struct Continuation : public TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, size_t n)
+      : k_(std::move(k)),
+        N_(n) {}
+
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      data_ = std::vector<int>(N_);
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      if (i_ == data_.size()) {
+        i_ = 0;
+        filled_ = true;
+      }
+      if (filled_) {
+        for (size_t j = 0; j < data_.size() - 1; ++j) {
+          data_[j] = data_[j + 1];
+        }
+        SetValue(data_.size() - 1, std::forward<Args>(args)...);
+      } else {
+        SetValue(i_++, std::forward<Args>(args)...);
+      }
+      stream_->Next();
+    }
+
+    void Ended() {
+      if (endedTriggered_) {
+        k_.Ended();
+      } else {
+        i_ = 0;
+        endedTriggered_ = true;
+        k_.Body(data_[i_++]);
+      }
+    }
+
+    void Next() override {
+      if (!endedTriggered_) {
+        stream_->Next();
+      } else {
+        if (i_ == data_.size()) {
+          k_.Ended();
+        } else {
+          k_.Body(data_[i_++]);
+        }
+      }
+    }
+
+    void Done() override {
+      k_.Ended();
+    }
+
+    void SetValue(size_t pos, int v) {
+      data_[pos] = v;
+    }
+
+    K_ k_;
+    size_t N_;
+    size_t i_ = 0;
+    bool filled_ = false;
+    bool endedTriggered_ = false;
+    std::vector<Arg_> data_;
+
+    TypeErasedStream* stream_ = nullptr;
+  };
+
+  struct Composable {
+    template <typename Arg>
+    using ValueFrom = Arg;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Arg>{std::move(k), std::move(N_)};
+    }
+
+    size_t N_;
+  };
+};
+
+////////////////////////////////////////////////////////////////////////
+
+struct _TakeRange {
+  template <typename K_, typename Arg_>
+  struct Continuation {
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      k_.Start(stream);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      if (filter_(std::forward<Args>(args)...)) {
+        inRange_ = true;
+        k_.Body(std::forward<Args>(args)...);
+      } else if (!inRange_) {
+        stream_->Next();
+      } else {
+        stream_->Done();
+      }
+    }
+
+    void Ended() {
+      k_.Ended();
+    }
+
+    bool filter_(Arg_ x) {
```

Comment:
Internal functions are usually still named e.g. `Filter()`, but you can make them `private` if you want to make it extra clear that they're internal. The naming of e.g. `filter_` is reserved for class variables.

Reply:
Why does this take `x`?